File: backend/app/api/v1/invoices.py
"""Tenant-scoped invoice endpoints (FEAT-008, TODO-075/076/078/079).

Base path: /api/v1/tenant/invoices
Guards: manage/invoices = admin+manager for writes; all staff can read.
"""

from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_admin_user, require_permission
from app.db.session import get_session
from app.models.admin_user import AdminUser
from app.models.enums import InvoiceStatus
from app.models.tenant import Tenant
from app.schemas.invoices import (
    InvoiceCreateRequest,
    InvoiceLineItemResponse,
    InvoiceListItem,
    InvoiceListResponse,
    InvoiceResponse,
    InvoiceUpdateRequest,
)
from app.schemas.transactions import (
    PaymentAllocationResponse,
    TransactionCreateRequest,
    TransactionResponse,
)
from app.services import invoices as invoice_service
from app.services import pdf as pdf_service
from app.services import transactions as transaction_service

router = APIRouter(prefix="/tenant/invoices", tags=["invoices"])


def _get_tenant_id(user: AdminUser) -> uuid.UUID:
    if user.tenant_id is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User must belong to a tenant",
        )
    return user.tenant_id


def _parse_uuid(value: str, *, kind: str) -> uuid.UUID:
    try:
        return uuid.UUID(value)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"{kind} not found",
        ) from exc


def _to_line_item_response(item: Any) -> InvoiceLineItemResponse:
    return InvoiceLineItemResponse(
        id=item.id,
        description=item.description,
        quantity=f"{item.quantity:.2f}",
        unit_price=f"{item.unit_price:.2f}",
        amount=f"{item.amount:.2f}",
        service_id=item.service_id,
        project_service_id=item.project_service_id,
    )


def _to_response(invoice: Any) -> InvoiceResponse:
    items = sorted(invoice.line_items, key=lambda li: li.created_at)
    return InvoiceResponse(
        id=invoice.id,
        invoice_number=invoice.invoice_number,
        status=invoice.status,
        project_id=invoice.project_id,
        client_id=invoice.project.client_id,
        issue_date=invoice.issue_date,
        due_date=invoice.due_date,
        subtotal=f"{invoice.subtotal:.2f}",
        tax_total=f"{invoice.tax_total:.2f}",
        total=f"{invoice.total:.2f}",
        notes=invoice.notes,
        line_items=[_to_line_item_response(li) for li in items],
        created_at=invoice.created_at,
        updated_at=invoice.updated_at,
    )


def _to_transaction_response(tx: Any) -> TransactionResponse:
    allocations = sorted(tx.allocations, key=lambda a: a.created_at)
    return TransactionResponse(
        id=tx.id,
        invoice_id=tx.invoice_id,
        amount=f"{tx.amount:.2f}",
        method=tx.method,
        reference_note=tx.reference_note,
        recorded_by_id=tx.recorded_by_id,
        recorded_at=tx.recorded_at,
        allocations=[
            PaymentAllocationResponse(
                id=a.id,
                line_item_id=a.line_item_id,
                amount=f"{a.amount:.2f}",
            )
            for a in allocations
        ],
    )


# ═══════════════════════════════════════════════════════════════════════════
# CRUD
# ═══════════════════════════════════════════════════════════════════════════


@router.post(
    "/",
    response_model=InvoiceResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_invoice_endpoint(
    body: InvoiceCreateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> InvoiceResponse:
    """Create a draft invoice. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    invoice = await invoice_service.create_draft_invoice(
        session,
        tenant_id=tenant_id,
        project_id=body.project_id,
        issue_date=body.issue_date,
        due_date=body.due_date,
        notes=body.notes,
        line_items=body.line_items,
        actor_id=user.id,
    )
    return _to_response(invoice)


@router.get("/", response_model=InvoiceListResponse)
async def list_invoices_endpoint(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    status_val: str | None = Query(default=None, alias="status"),
    project_id: str | None = Query(default=None),
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> InvoiceListResponse:
    """List invoices for a tenant. All staff can read."""
    tenant_id = _get_tenant_id(user)

    status_filter: InvoiceStatus | None = None
    if status_val:
        try:
            status_filter = InvoiceStatus(status_val)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                detail=(
                    f"Invalid status: {status_val}. "
                    "Must be 'draft', 'issued', 'partially_paid', 'paid', or 'void'."
                ),
            ) from None

    parsed_project_id: uuid.UUID | None = None
    if project_id:
        try:
            parsed_project_id = uuid.UUID(project_id)
        except ValueError as exc:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                detail="project_id must be a valid UUID",
            ) from exc

    result = await invoice_service.list_invoices(
        session,
        tenant_id=tenant_id,
        page=page,
        page_size=page_size,
        status_filter=status_filter,
        project_id=parsed_project_id,
    )
    items = [InvoiceListItem(**item) for item in result["items"]]
    return InvoiceListResponse(
        items=items,
        total=result["total"],
        page=result["page"],
        page_size=result["page_size"],
    )


@router.get("/{invoice_id}", response_model=InvoiceResponse)
async def get_invoice_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> InvoiceResponse:
    """Get invoice detail with line items. All staff can read."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    invoice = await invoice_service.get_invoice(session, tenant_id=tenant_id, invoice_id=iid)
    return _to_response(invoice)


@router.get("/{invoice_id}/pdf")
async def get_invoice_pdf_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> Response:
    """Render an invoice as a PDF attachment (TODO-084/085). All staff can read."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    invoice = await invoice_service.get_invoice(session, tenant_id=tenant_id, invoice_id=iid)
    tenant = await session.get(Tenant, tenant_id)
    pdf_bytes = await pdf_service.render_invoice_pdf(session, invoice=invoice, tenant=tenant)
    filename = invoice.invoice_number or "DRAFT"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{filename}.pdf"'},
    )


@router.patch("/{invoice_id}", response_model=InvoiceResponse)
async def update_invoice_endpoint(
    invoice_id: str,
    body: InvoiceUpdateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> InvoiceResponse:
    """Update a draft invoice (or notes on an issued one). Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    invoice = await invoice_service.update_draft_invoice(
        session,
        tenant_id=tenant_id,
        invoice_id=iid,
        issue_date=body.issue_date,
        due_date=body.due_date,
        notes=body.notes,
        line_items=body.line_items,
        actor_id=user.id,
    )
    return _to_response(invoice)


@router.delete(
    "/{invoice_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_invoice_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> Response:
    """Delete a draft invoice. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    try:
        await invoice_service.delete_draft_invoice(
            session,
            tenant_id=tenant_id,
            invoice_id=iid,
            actor_id=user.id,
        )
    except invoice_service.InvoiceDeleteNotAllowedError:
        raise HTTPException(
            status_code=status.HTTP_405_METHOD_NOT_ALLOWED,
            detail="Issued invoices cannot be deleted",
        ) from None
    return Response(status_code=status.HTTP_204_NO_CONTENT)


# ═══════════════════════════════════════════════════════════════════════════
# Issue (TODO-078)
# ═══════════════════════════════════════════════════════════════════════════


@router.post("/{invoice_id}/issue", response_model=InvoiceResponse)
async def issue_invoice_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> InvoiceResponse:
    """Issue a draft invoice: assign number + issue_date. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    invoice = await invoice_service.issue_invoice(
        session, tenant_id=tenant_id, invoice_id=iid, actor_id=user.id
    )
    return _to_response(invoice)


@router.post("/{invoice_id}/void", response_model=InvoiceResponse)
async def void_invoice_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> InvoiceResponse:
    """Void an issued/partially-paid/paid invoice. Number + data retained. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    invoice = await invoice_service.void_invoice(
        session, tenant_id=tenant_id, invoice_id=iid, actor_id=user.id
    )
    return _to_response(invoice)


# ═══════════════════════════════════════════════════════════════════════════
# Transactions (FEAT-009, TODO-089/090/092/093/094)
# ═══════════════════════════════════════════════════════════════════════════


@router.post(
    "/{invoice_id}/transactions",
    response_model=TransactionResponse,
    status_code=status.HTTP_201_CREATED,
)
async def record_transaction_endpoint(
    invoice_id: str,
    body: TransactionCreateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "invoices")),
) -> TransactionResponse:
    """Record a payment against an issued/partially-paid invoice. Admin/Manager only."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    tx = await transaction_service.record_transaction(
        session,
        tenant_id=tenant_id,
        invoice_id=iid,
        amount=body.amount,
        method=body.method,
        reference_note=body.reference_note,
        recorded_at=body.recorded_at,
        allocations=body.allocations,
        actor_id=user.id,
    )
    return _to_transaction_response(tx)


@router.get(
    "/{invoice_id}/transactions",
    response_model=list[TransactionResponse],
)
async def list_transactions_endpoint(
    invoice_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> list[TransactionResponse]:
    """List payments recorded against an invoice. All staff can read."""
    tenant_id = _get_tenant_id(user)
    iid = _parse_uuid(invoice_id, kind="Invoice")
    txs = await transaction_service.list_transactions(session, tenant_id=tenant_id, invoice_id=iid)
    return [_to_transaction_response(tx) for tx in txs]

File: frontend/src/lib/api/invoices.js
import { ApiError, apiFetch, BASE_URL } from './client.js';

/**
 * Tenant invoicing API endpoints (FR-7.6).
 *
 * Accessible to all staff (admin, manager, employee) for reads and
 * create/update on draft invoices; issuing and voiding are server-enforced.
 */

/**
 * @typedef {object} InvoiceListItem
 * @property {string} id
 * @property {string|null} invoice_number null while draft
 * @property {'draft'|'issued'|'partially_paid'|'paid'|'void'} status
 * @property {string|null} project_id null for general (internal) invoices
 * @property {string|null} client_id null for general (internal) invoices
 * @property {string|null} issue_date ISO date
 * @property {string|null} due_date ISO date
 * @property {number|string} total decimal-as-string or number
 * @property {string} created_at
 */

/**
 * @typedef {object} InvoiceLineItem
 * @property {string} id
 * @property {string} description
 * @property {number} quantity
 * @property {number|string} unit_price decimal-as-string or number
 * @property {number|string} amount decimal-as-string or number
 * @property {string|null} entry_date ISO date the item was delivered/performed; null when unset
 * @property {string|null} service_id
 * @property {string|null} project_service_id null for custom lines
 */

/**
 * @typedef {object} InvoiceDetailResponse
 * @property {string} id
 * @property {string|null} invoice_number null while draft
 * @property {'draft'|'issued'|'partially_paid'|'paid'|'void'} status
 * @property {string|null} project_id null for general (internal) invoices
 * @property {string|null} client_id null for general (internal) invoices
 * @property {boolean} is_general true when project_id/client_id are null
 * @property {string|null} issue_date
 * @property {string|null} due_date
 * @property {number|string} subtotal
 * @property {number|string} tax_total
 * @property {number|string} total
 * @property {string|null} notes
 * @property {InvoiceLineItem[]} line_items
 * @property {string} created_at
 * @property {string} updated_at
 */

/**
 * @typedef {object} ProjectOverviewResponse
 * @property {string} project_id
 * @property {string} name
 * @property {string} status
 * @property {number} milestone_total
 * @property {number} milestone_completed
 * @property {number} milestone_completion_pct
 * @property {string} total_invoiced decimal-as-string
 * @property {string} total_paid decimal-as-string
 * @property {string} balance_due decimal-as-string
 * @property {Array<{ id: string, number: string|null, status: string, total: string }>} linked_invoices
 * @property {Array<{ service_id: string, service_name: string, total_invoiced: string, total_paid: string, total_outstanding: string }>} service_breakdown
 */

/**
 * @typedef {object} TransactionAllocation
 * @property {string} id
 * @property {string} line_item_id
 * @property {string} amount decimal-as-string
 */

/**
 * @typedef {object} Transaction
 * @property {string} id
 * @property {string} invoice_id
 * @property {string} amount decimal-as-string
 * @property {'debit'|'credit'} direction debit = payment (money in), credit = refund (money out)
 * @property {'bank_transfer'|'card'|'cash'|'other'} method
 * @property {string} reference_note
 * @property {string|null} recorded_by_id
 * @property {string} recorded_at ISO datetime
 * @property {TransactionAllocation[]} allocations
 */

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {{ page?: number, page_size?: number, status?: string, project_id?: string, client_id?: string }} [params]
 *   `client_id` filters to that client's project invoices (general/internal excluded server-side).
 * @returns {Promise<{ items: InvoiceListItem[], total: number, page: number, page_size: number }>}
 */
export function listInvoices(fetchFn, token, params = {}) {
	return apiFetch(fetchFn, '/tenant/invoices', { token, params });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<InvoiceDetailResponse>}
 */
export function getInvoice(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(id)}`, { token });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {{
 *   project_id?: string,
 *   issue_date?: string,
 *   due_date?: string,
 *   notes?: string,
 *   line_items: Array<{ project_service_id?: string, description?: string, unit_price?: string|number, quantity?: number, entry_date?: string }>
 * }} body Create payload; omit project_id for general (internal) invoices
 * @returns {Promise<InvoiceDetailResponse>}
 */
export function createInvoice(fetchFn, token, body) {
	return apiFetch(fetchFn, '/tenant/invoices', { method: 'POST', token, body });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @param {Record<string, any>} body partial fields (issue_date, due_date, notes, line_items)
 * @returns {Promise<InvoiceDetailResponse>}
 */
export function updateInvoice(fetchFn, token, id, body) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(id)}`, {
		method: 'PATCH',
		token,
		body
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<null>} 204 on success
 */
export function deleteInvoice(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(id)}`, {
		method: 'DELETE',
		token
	});
}

/**
 * Record a payment against an invoice. `allocations` null/omitted means the
 * backend auto-allocates proportionally; pass it to override manually.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @param {{
 *   amount: string,
 *   method: 'bank_transfer'|'card'|'cash'|'other',
 *   reference_note?: string,
 *   recorded_at?: string,
 *   allocations?: Array<{ line_item_id: string, amount: string }>
 * }} payload
 * @returns {Promise<Transaction>}
 */
export function recordTransaction(fetchFn, token, invoiceId, payload) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(invoiceId)}/transactions`, {
		method: 'POST',
		token,
		body: payload
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @returns {Promise<Transaction[]>}
 */
export function listTransactions(fetchFn, token, invoiceId) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(invoiceId)}/transactions`, {
		token
	});
}

/**
 * Record a refund (credit transaction) against an invoice. Cannot exceed the
 * invoice's net paid amount (server-enforced).
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @param {{
 *   amount: string,
 *   method?: 'bank_transfer'|'card'|'cash'|'other',
 *   reference_note?: string
 * }} payload
 * @returns {Promise<Transaction>}
 */
export function refundInvoice(fetchFn, token, invoiceId, payload) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(invoiceId)}/refund`, {
		method: 'POST',
		token,
		body: payload
	});
}

/**
 * Apply client advance balance to an issued/partially-paid invoice. Omitting
 * `amount` applies as much as the advance balance + invoice balance allow.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @param {string} [amount] decimal-as-string, optional
 * @returns {Promise<{ applied: string, advance_balance: string }>}
 */
export function applyAdvance(fetchFn, token, invoiceId, amount) {
	/** @type {Record<string, any>} */
	const body = {};
	if (amount != null && amount !== '') body.amount = amount;
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(invoiceId)}/apply-advance`, {
		method: 'POST',
		token,
		body
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<InvoiceDetailResponse>}
 */
export function issueInvoice(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(id)}/issue`, {
		method: 'POST',
		token
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} id
 * @returns {Promise<InvoiceDetailResponse>}
 */
export function voidInvoice(fetchFn, token, id) {
	return apiFetch(fetchFn, `/tenant/invoices/${encodeURIComponent(id)}/void`, {
		method: 'POST',
		token
	});
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} projectId
 * @returns {Promise<ProjectOverviewResponse>}
 */
export function getProjectOverview(fetchFn, token, projectId) {
	return apiFetch(fetchFn, `/tenant/projects/${encodeURIComponent(projectId)}/overview`, { token });
}

/**
 * Fetch a raw binary payload with a Bearer token.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} path API path starting with '/', relative to /api/v1
 * @param {string} token
 * @returns {Promise<Blob>}
 * @throws {ApiError}
 */
async function fetchBlob(fetchFn, path, token) {
	const res = await fetchFn(`${BASE_URL}${path}`, {
		headers: { Authorization: `Bearer ${token}` }
	});
	if (!res.ok) {
		let data = null;
		try {
			data = await res.json();
		} catch {
			// non-JSON error body; fall through to generic error
		}
		const envelope = data && data.error ? data.error : {};
		throw new ApiError(
			res.status,
			envelope.code ?? 'UNKNOWN',
			envelope.message ?? res.statusText,
			envelope.details ?? {}
		);
	}
	return res.blob();
}

/**
 * Download an invoice PDF attachment. Must run in the browser (uses
 * `document`); call from an event handler, not a load function.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @param {string} filename e.g. "INV-0001.pdf"
 * @returns {Promise<void>}
 * @throws {ApiError}
 */
export async function downloadInvoicePdf(fetchFn, token, invoiceId, filename) {
	const blob = await fetchBlob(
		fetchFn,
		`/tenant/invoices/${encodeURIComponent(invoiceId)}/pdf`,
		token
	);
	const url = URL.createObjectURL(blob);
	const a = document.createElement('a');
	a.href = url;
	a.download = filename;
	document.body.appendChild(a);
	a.click();
	a.remove();
	URL.revokeObjectURL(url);
}

/**
 * Open an invoice PDF in a new tab. The PDF endpoint requires an Authorization
 * header, so the blob is fetched first and opened via an object URL (revoked
 * after 60s so the new tab keeps the loaded document).
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} invoiceId
 * @param {string} filename e.g. "INV-0001.pdf"
 * @returns {Promise<void>}
 * @throws {ApiError}
 */
// eslint-disable-next-line no-unused-vars -- filename kept for parity with the download helper
export async function viewInvoicePdf(fetchFn, token, invoiceId, filename) {
	const blob = await fetchBlob(
		fetchFn,
		`/tenant/invoices/${encodeURIComponent(invoiceId)}/pdf`,
		token
	);
	const url = URL.createObjectURL(blob);
	window.open(url, '_blank');
	setTimeout(() => URL.revokeObjectURL(url), 60000);
}
